tools/storage-walrus/src/read_json.rs

```diff
@@ -0,0 +1,610 @@
+                        // If it's not valid JSON, return an error using ReadJsonError::InvalidJson
+                        return Output::Err {
+                            reason: ReadJsonError::InvalidJson(e.to_string()).to_string(),
+                            kind: ReadErrorKind::Validation,
+                            status_code: None,
+                        };
+                    }
+                };
+
+                // If a JSON schema was provided, validate against it
+                if let Some(schema_def) = input.json_schema.as_ref() {
+                    // Create a validation configuration using all schema fields
+                    let validation_result =
+                        match create_validation_configuration(schema_def, &json_data) {
+                            Ok(result) => result,
+                            Err(e) => {
+                                return Output::Err {
+                                    reason: e.to_string(),
+                                    kind: ReadErrorKind::Schema,
+                                    status_code: None,
+                                };
+                            }
+                        };
+
+                    if validation_result.is_valid {
+                        Output::Ok { json: json_data }
+                    } else {
+                        Output::Err {
+                            reason: ReadJsonError::ValidationError(validation_result.error_message)
+                                .to_string(),
+                            kind: ReadErrorKind::Schema,
+                            status_code: None,
+                        }
+                    }
+                } else {
+                    // If we parsed valid JSON but no schema was provided
+                    Output::Ok { json: json_data }
+                }
+            }
+            Err(e) => {
+                // Extract status code from WalrusError if available
+                let status_code = match &e {
+                    ReadJsonError::ReadError(WalrusError::ApiError { status_code, .. }) => {
+                        Some(*status_code)
+                    }
+                    _ => None,
+                };
+
+                Output::Err {
+                    reason: e.to_string(),
+                    kind: ReadErrorKind::Network,
+                    status_code,
+                }
+            }
+        }
+    }
+}
+
+impl ReadJson {
+    async fn read(
+        &self,
+        blob_id: String,
+        aggregator_url: Option<String>,
+    ) -> Result<String, ReadJsonError> {
+        let walrus_client = WalrusConfig::new()
+            .with_aggregator_url(aggregator_url)
+            .build();
+
+        let storage_info = walrus_client.read_json(&blob_id).await?;
+
+        Ok(storage_info)
+    }
+}
+
+// Validation utility struct and function to use all schema fields
+struct ValidationResult {
+    is_valid: bool,
+    error_message: String,
+}
```

Comment:
The `ValidationResult` struct introduces ambiguity - it's possible to have `is_valid` set to true while still containing an error message, which can be misleading.

A more idiomatic approach is to have `create_validation_configuration` return a `Result<(), ReadJsonError>`, so you avoid some of the boolean checks you have later.